# Arm C: Data Driven Clusters (k prototypes)

This notebook implements Arm C of the study. Within each outer training fold, patients are grouped by an unsupervised clustering algorithm, and each classifier is tuned and trained separately within each cluster. Arm C answers RQ2 (does data driven stratification change performance compared with the global model of Arm A?) and, together with Arm B, RQ3.

k prototypes (Huang, 1998) is used because the features are a mix of continuous and categorical variables. It measures distance with squared Euclidean distance on the numeric features plus a weighted count of mismatches on the categorical ones, whereas k means suits only continuous data and k modes only categorical data (proposal, Section 3.3).

Everything except the grouping is held identical to Arm A by importing it from `study_config.py`: the dataset, the 9 features, the preprocessing, the three classifiers and their grids, the inner tuning procedure, the outer folds (`fold_id.csv`, created by Arm A) and the 0.5 threshold. Unlike Arm B, all 9 features stay in the classifiers, because no feature is constant within a cluster.

## 1. Imports and configuration

All shared settings come from `study_config.py`. Three settings are specific to the clustering step and are fixed before looking at any results:

* `K_CLUSTER_CANDIDATES = [2, 3, 4]`: the cluster counts considered in each training fold. The proposal requires a small number of clusters given about 300 patients (Section 3.3).
* `MIN_CLUSTER_SIZE = 40`: a candidate number of clusters is rejected if any cluster would have fewer training patients than this.
* `N_INIT_KPROTO = 10`: number of random restarts of k prototypes per candidate, the library default.

In [1]:
import hashlib
import json
import os

import numpy as np
import pandas as pd

from sklearn.model_selection import StratifiedKFold, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, silhouette_score, adjusted_rand_score
from kmodes.kprototypes import KPrototypes

import study_config as cfg
from study_config import RANDOM_STATE, K_OUTER, K_INNER, THRESHOLD, FEATURE_SET

cfg.make_dirs()
K_CLUSTER_CANDIDATES = [2, 3, 4]
MIN_CLUSTER_SIZE = 40
N_INIT_KPROTO = 10
print(f"FEATURE_SET={FEATURE_SET!r}, K_OUTER={K_OUTER}, K_INNER={K_INNER}, "
      f"THRESHOLD={THRESHOLD}, RANDOM_STATE={RANDOM_STATE}, tuning={cfg.TUNING_SCORING!r}")

FEATURE_SET='selected', K_OUTER=10, K_INNER=10, THRESHOLD=0.5, RANDOM_STATE=2, tuning='roc_auc'


## 2. Load the dataset and build the target

Arm C uses the same cleaned file and the same binary target (`num > 0` means disease present) as Arms A and B.

In [2]:
df = pd.read_csv(cfg.CLEAN_DATA_FILE)
assert df.shape[0] == 297, "Arm C expects the same 297 patient dataset as Arm A."
y = (df["num"] > 0).astype(int)
print("Loaded shape:", df.shape, "| positive rate:", round(y.mean(), 3))

Loaded shape: (297, 15) | positive rate: 0.461


## 3. Features for the classifiers and for the clustering

The same 9 features are used for two different purposes, with two different encodings:

* **Classifiers.** Exactly as in Arm A: `study_config.get_preprocessor()` one hot encodes the nominal features and standardises every column, inside each pipeline.
* **Clustering.** k prototypes needs a numeric block and a categorical block. The four continuous features and the two binary features (`sex`, `exang`) form the numeric block and are standardised together with one `StandardScaler` fitted on the training fold. The three nominal features (`cp`, `slope`, `thal`) form the categorical block and are passed as raw codes, compared by simple matching. One hot encoding them first would turn each category into several numeric columns under Euclidean distance, which is exactly what choosing k prototypes over k means is meant to avoid.

The binary features are standardised for clustering for the same reason they are standardised for the classifiers: squared Euclidean distance weights each numeric feature by its spread, so a raw 0/1 column next to standardised continuous columns would count for much less. Section 4 checks how much this choice changes the clusters.

In [3]:
X = df[cfg.get_feature_list(FEATURE_SET)]
groups = cfg.get_feature_groups(FEATURE_SET)
continuous, nominal, binary = groups["continuous"], groups["nominal"], groups["binary"]

cluster_numeric_cols = continuous + binary
cluster_categorical_idx = list(range(len(cluster_numeric_cols), len(cluster_numeric_cols) + len(nominal)))
print(f"Features used ({X.shape[1]}): {list(X.columns)}")
print("Clustering numeric block:", cluster_numeric_cols)
print("Clustering categorical block:", nominal)

Features used (9): ['age', 'thalach', 'oldpeak', 'ca', 'cp', 'slope', 'thal', 'sex', 'exang']
Clustering numeric block: ['age', 'thalach', 'oldpeak', 'ca', 'sex', 'exang']
Clustering categorical block: ['cp', 'slope', 'thal']


## 4. Clustering helpers and a descriptive whole dataset fit

The helper functions below are used unchanged by the leakage safe per fold procedure in Section 6:

* `build_cluster_arrays` builds the numeric and categorical arrays, fitting the scaler on the data it is given (a training fold) or applying an already fitted scaler (a validation fold).
* `fit_kprototypes_select_k` fits k prototypes for each candidate number of clusters on predictor variables only. The outcome is never passed in, so it cannot influence the clustering. Candidates that create a cluster smaller than `MIN_CLUSTER_SIZE` are rejected, and among the rest the one with the highest silhouette score is chosen (proposal, Section 3.3). The silhouette is computed with the same mixed distance k prototypes minimises, so the selection criterion matches the clustering objective.
* `predict_clusters` assigns new patients to the nearest already fitted cluster centre. It never refits.

Before the evaluation, k prototypes is fitted once on all 297 patients **purely to describe** what it finds. This descriptive fit is not used anywhere in the evaluation, where clustering is refitted inside each training fold.

In [4]:
def build_cluster_arrays(frame, scaler=None, fit_scaler=False, scale_binary=True):
    # scale_binary=False leaves the binary features on their raw 0/1 scale (sensitivity check only).
    numeric = continuous + binary if scale_binary else continuous
    if fit_scaler:
        scaler = StandardScaler().fit(frame[numeric])
    Xnum = scaler.transform(frame[numeric])
    if not scale_binary:
        Xnum = np.hstack([Xnum, frame[binary].to_numpy(dtype=float)])
    return Xnum, frame[nominal].to_numpy(), scaler


def mixed_distance_matrix(Xnum_a, Xcat_a, Xnum_b, Xcat_b, gamma):
    # The k prototypes cost: squared Euclidean on the numeric block plus gamma
    # times the number of categorical mismatches.
    num_sq = ((Xnum_a[:, None, :] - Xnum_b[None, :, :]) ** 2).sum(axis=2)
    cat_mismatch = (Xcat_a[:, None, :] != Xcat_b[None, :, :]).sum(axis=2)
    return num_sq + gamma * cat_mismatch


def fit_kprototypes_select_k(frame, scale_binary=True):
    # Predictors only: the outcome is never an argument.
    Xnum, Xcat, scaler = build_cluster_arrays(frame, fit_scaler=True, scale_binary=scale_binary)
    best = None
    for k in K_CLUSTER_CANDIDATES:
        if len(frame) < MIN_CLUSTER_SIZE * k:
            continue
        model = KPrototypes(n_clusters=k, init="Cao", n_init=N_INIT_KPROTO, random_state=RANDOM_STATE, verbose=0)
        labels = model.fit_predict(np.hstack([Xnum, Xcat]), categorical=cluster_categorical_idx)
        if np.bincount(labels, minlength=k).min() < MIN_CLUSTER_SIZE:
            continue
        sil = silhouette_score(mixed_distance_matrix(Xnum, Xcat, Xnum, Xcat, model.gamma), labels, metric="precomputed")
        if best is None or sil > best["silhouette"]:
            best = {"model": model, "scaler": scaler, "labels": labels, "k": k, "silhouette": sil}
    if best is None:
        return None, None, None, None, None
    return best["model"], best["scaler"], best["labels"], best["k"], best["silhouette"]


def predict_clusters(model, scaler, frame):
    Xnum, Xcat, _ = build_cluster_arrays(frame, scaler=scaler)
    return model.predict(np.hstack([Xnum, Xcat]), categorical=cluster_categorical_idx)


desc_model, desc_scaler, desc_labels, desc_k, desc_sil = fit_kprototypes_select_k(X)
print(f"Whole dataset descriptive clustering selected k={desc_k} (silhouette={desc_sil:.3f})")
desc = pd.DataFrame({"cluster": desc_labels, "sex": df["sex"].map({1.0: "male", 0.0: "female"}), "disease": y})
print(f"Adjusted Rand index vs outcome: {adjusted_rand_score(y, desc_labels):.3f}")
print(f"Adjusted Rand index vs sex:     {adjusted_rand_score(df['sex'], desc_labels):.3f}")
pd.crosstab(desc["cluster"], desc["sex"], margins=True)

Whole dataset descriptive clustering selected k=2 (silhouette=0.391)
Adjusted Rand index vs outcome: 0.369
Adjusted Rand index vs sex:     0.010


sex,female,male,All
cluster,,,
0,28,99,127
1,68,102,170
All,96,201,297


**Sensitivity check on scaling.** The whole dataset clustering is repeated with the binary features left on their raw 0/1 scale, to see whether that encoding choice changes the clusters or their relationship with sex and with the outcome.

In [5]:
sens_rows = []
for label, scale_bin in [("binary features raw 0/1", False), ("binary features standardised (used)", True)]:
    m, s, lab, k_s, sil_s = fit_kprototypes_select_k(X, scale_binary=scale_bin)
    sens_rows.append({
        "scaling": label, "k": k_s, "silhouette": round(sil_s, 3),
        "cluster_sizes": np.bincount(lab, minlength=k_s).tolist(),
        "disease_rates": [round(y[lab == c].mean(), 3) for c in range(k_s)],
        "pct_male": [round((df["sex"][lab == c] == 1.0).mean() * 100, 1) for c in range(k_s)],
        "ari_vs_outcome": round(adjusted_rand_score(y, lab), 3),
        "ari_vs_sex": round(adjusted_rand_score(df["sex"], lab), 3),
    })
sensitivity_df = pd.DataFrame(sens_rows)
sensitivity_df.to_csv(os.path.join(cfg.DIAGNOSTICS_DIR, "armC_scaling_sensitivity.csv"), index=False)
sensitivity_df

,scaling,k,silhouette,cluster_sizes,disease_rates,pct_male,ari_vs_outcome,ari_vs_sex
0,binary features raw 0/1,2,0.433,"[123, 174]","[0.78, 0.236]","[74.8, 62.6]",0.291,-0.003
1,binary features standardised (used),2,0.391,"[127, 170]","[0.811, 0.2]","[78.0, 60.0]",0.369,0.010


**Silhouette by number of clusters.** As a check on how much cluster structure the data contain, the silhouette is computed for every k from 2 to 6 on the whole dataset. As a rough guide, values below about 0.25 indicate weak or no real structure and 0.25 to 0.5 weak to moderate structure. Silhouette also tends to favour small k, so its preference for few clusters should not be over read. Because the distance is squared Euclidean (not a true metric), these values are comparable across k here but not with silhouette values reported elsewhere.

In [6]:
Xnum_full, Xcat_full, _ = build_cluster_arrays(X, fit_scaler=True)
sil_by_k = {}
for k in range(2, 7):
    m = KPrototypes(n_clusters=k, init="Cao", n_init=N_INIT_KPROTO, random_state=RANDOM_STATE, verbose=0)
    lab = m.fit_predict(np.hstack([Xnum_full, Xcat_full]), categorical=cluster_categorical_idx)
    sil_by_k[k] = silhouette_score(mixed_distance_matrix(Xnum_full, Xcat_full, Xnum_full, Xcat_full, m.gamma),
                                   lab, metric="precomputed")
sil_df = pd.Series(sil_by_k, name="silhouette").rename_axis("k").reset_index()
sil_df.to_csv(os.path.join(cfg.DIAGNOSTICS_DIR, "armC_silhouette_by_k.csv"), index=False)
sil_df.round(3)

,k,silhouette
0,2,0.391
1,3,0.356
2,4,0.370
3,5,0.329
4,6,0.319


**What the descriptive clusters look like.** The table below profiles each whole dataset cluster on the outcome, sex and the main clinical features, to see what kind of patient grouping k prototypes recovers.

In [7]:
characterization = pd.DataFrame({
    "n": desc.groupby("cluster").size(),
    "disease_rate": desc.groupby("cluster")["disease"].mean(),
    "pct_male": desc.groupby("cluster")["sex"].apply(lambda s: (s == "male").mean() * 100),
    "mean_age": X.groupby(desc_labels)["age"].mean(),
    "mean_thalach": X.groupby(desc_labels)["thalach"].mean(),
    "pct_exang": X.groupby(desc_labels)["exang"].mean() * 100,
    "mean_oldpeak": X.groupby(desc_labels)["oldpeak"].mean(),
    "mean_ca": X.groupby(desc_labels)["ca"].mean(),
    "pct_cp_asymptomatic": X.groupby(desc_labels)["cp"].apply(lambda s: (s == 4.0).mean() * 100),
}).round(2)
characterization.to_csv(os.path.join(cfg.DIAGNOSTICS_DIR, "armC_cluster_characterization.csv"))
characterization

,n,disease_rate,pct_male,mean_age,mean_thalach,pct_exang,mean_oldpeak,mean_ca,pct_cp_asymptomatic
0,127,0.81,77.95,58.77,132.67,63.78,1.87,1.22,74.80
1,170,0.20,60.00,51.38,162.25,9.41,0.45,0.27,27.65


**Reading the descriptive clustering.** On all 297 patients k prototypes chooses two clusters (silhouette 0.39), and they line up with **disease severity**, not with sex:

* cluster 0 (127 patients) is the high risk group: 81% disease, older (mean age 58.8), lower maximum heart rate (133), 64% exercise angina, mean ST depression 1.87, mean vessel count 1.22, 75% asymptomatic chest pain;
* cluster 1 (170 patients) is the low risk group: 20% disease, younger (51.4), higher maximum heart rate (162), 9% exercise angina, mean ST depression 0.45, mean vessel count 0.27, 28% asymptomatic chest pain.

The adjusted Rand index between the clusters and sex is essentially zero (0.01), so the data driven grouping recovers nothing like the clinical grouping used in Arm B. Its agreement with the outcome is moderate (adjusted Rand index 0.37). Leaving the binary features on their raw scale gives almost the same two clusters (123 and 174 patients, 78% and 24% disease) and the same conclusion: the clusters follow severity and ignore sex.

The silhouette values (0.32 to 0.39 for k from 2 to 6, highest at k = 2) indicate weak to moderate structure. The data contain a real severity gradient, but not well separated patient types. That matters for interpreting this arm: splitting along a gradient the classifier could already learn from the raw features gives each cluster model less data without obviously giving it new information.

## 5. Reuse Arm A's outer folds

Arm C loads `fold_id.csv` from Arm A, so every patient is held out in exactly the same fold in every arm. The file's MD5 hash and the configuration recorded in `run_manifest.json` are checked against this notebook's own settings first; any mismatch stops the notebook.

In [8]:
fold_id = pd.read_csv(cfg.FOLD_FILE)["fold"].to_numpy()
with open(cfg.MANIFEST_FILE) as f:
    manifest = json.load(f)
assert hashlib.md5(open(cfg.FOLD_FILE, "rb").read()).hexdigest() == manifest["fold_id_md5"], \
    "fold_id.csv changed since Arm A ran; rerun Arm A, then this notebook."
for key, value in {"feature_set": FEATURE_SET, "k_outer": K_OUTER, "k_inner": K_INNER,
                   "random_state": RANDOM_STATE, "threshold": THRESHOLD,
                   "tuning_scoring": cfg.TUNING_SCORING}.items():
    assert manifest[key] == value, f"run_manifest.json {key}={manifest[key]!r} but this notebook uses {value!r}"
assert len(fold_id) == len(df) and set(np.unique(fold_id)) == set(range(K_OUTER))
print("fold_id.csv and run_manifest.json verified against study_config.")

fold_id.csv and run_manifest.json verified against study_config.


## 6. Per fold clustering

Within each outer fold, the scaler and k prototypes are fitted on that fold's **training patients only**, and the validation patients are assigned to the nearest fitted cluster centre without refitting. Fitting the clustering on all patients would let the validation patients shape the clusters their own model is chosen by, which is a leak (proposal, Section 3.5).

Two stages are applied, both fixed before any results were seen:

1. **Choosing k (outcome blind).** `fit_kprototypes_select_k` rejects any k that creates a cluster under 40 patients and picks the remaining k with the highest silhouette. The outcome plays no part.
2. **Feasibility check (after k is fixed).** Every chosen cluster must contain at least `K_INNER` training patients of each class, so that the inner stratified 10 fold tuning is well defined inside it. This uses training labels only and never changes k.

If no k passes stage 1, or the chosen k fails stage 2, that fold falls back to a single cluster, which makes it identical in procedure to Arm A for that fold. The reason is recorded in `diagnostics/armC_cluster_choice.csv`.

In [9]:
def cluster_or_fallback(X_train, y_train):
    model, scaler, labels, k, sil = fit_kprototypes_select_k(X_train)
    if model is None:
        return np.zeros(len(X_train), dtype=int), None, None, 1, "no_k_passed_size_guard"
    y_arr = np.asarray(y_train)
    if min(np.bincount(y_arr[labels == c], minlength=2).min() for c in range(k)) < K_INNER:
        return np.zeros(len(X_train), dtype=int), None, None, 1, "chosen_k_failed_class_count"
    return labels, scaler, model, k, None


fold_clusters = {}
choice_rows = []
for k in range(K_OUTER):
    train, validation = fold_id != k, fold_id == k
    labels_train, scaler, model, k_used, fallback_reason = cluster_or_fallback(X[train], y[train])
    labels_val = np.zeros(validation.sum(), dtype=int) if model is None else predict_clusters(model, scaler, X[validation])
    fold_clusters[k] = (labels_train, labels_val, k_used)
    choice_rows.append({
        "fold": k, "k_used": k_used, "fallback_reason": fallback_reason,
        "train_cluster_sizes": np.bincount(labels_train, minlength=k_used).tolist(),
        "train_cluster_disease_rates": [round(y[train][labels_train == c].mean(), 2) for c in range(k_used)],
        "val_cluster_sizes": np.bincount(labels_val, minlength=k_used).tolist(),
    })
cluster_choice_df = pd.DataFrame(choice_rows)
cluster_choice_df.to_csv(os.path.join(cfg.DIAGNOSTICS_DIR, "armC_cluster_choice.csv"), index=False)
cluster_choice_df

,fold,k_used,fallback_reason,train_cluster_sizes,train_cluster_disease_rates,val_cluster_sizes
0,0,2,None,"[114, 153]","[0.82, 0.19]","[13, 17]"
1,1,2,None,"[114, 153]","[0.81, 0.21]","[12, 18]"
2,2,2,None,"[120, 147]","[0.81, 0.18]","[8, 22]"
3,3,2,None,"[113, 154]","[0.82, 0.19]","[14, 16]"
4,4,2,None,"[111, 156]","[0.82, 0.21]","[11, 19]"
5,5,2,None,"[115, 152]","[0.8, 0.2]","[14, 16]"
6,6,2,None,"[152, 115]","[0.2, 0.8]","[19, 11]"
7,7,2,None,"[112, 156]","[0.81, 0.21]","[12, 17]"
8,8,2,None,"[113, 155]","[0.82, 0.2]","[15, 14]"
9,9,2,None,"[113, 155]","[0.81, 0.2]","[14, 15]"


## 7. Nested cross validation within clusters

For each outer fold and classifier, a separate model is tuned and trained for each cluster, using `GridSearchCV` with an inner stratified 10 fold split of that cluster's training patients, the same grid and the same ROC AUC criterion as Arm A. The outer validation fold is never used for tuning.

The clustering is fitted once per outer fold (Section 6), not again inside every inner split. This means patients in an inner validation split helped shape the cluster they are scored within, so the inner score used to pick hyperparameters is slightly optimistic. It does not bias the outer estimate, which is the reported result, because outer validation patients never influence the clustering. Refitting the clustering inside every inner split would remove even that slight optimism at about ten times the cost, and could only change which hyperparameters are chosen.

As in Arm B, each validation patient is scored by their own cluster's model and classed with the shared 0.5 threshold, and the predictions from all clusters are pooled into one set covering the whole validation fold before accuracy, F1 and ROC AUC are computed. `roc_auc_within_group` records the patient weighted average of the per cluster AUCs, as a check on the pooled AUC (see the caveat in Arm B, Section 6).

In [10]:
models = cfg.get_models(FEATURE_SET)

fold_rows, cluster_rows, oof_rows, best_param_rows = [], [], [], []
patient_id = df.index.to_numpy()

for k in range(K_OUTER):
    train, validation = fold_id != k, fold_id == k
    labels_train, labels_val, k_used = fold_clusters[k]

    for name, (pipe, grid) in models.items():
        pooled_true, pooled_proba, pooled_pred, group_auc, group_n = [], [], [], [], []
        for c in range(k_used):
            c_train, c_val = train.copy(), validation.copy()
            c_train[train] = labels_train == c
            c_val[validation] = labels_val == c
            if c_val.sum() == 0:
                continue  # no validation patient was assigned to this cluster

            inner = StratifiedKFold(n_splits=K_INNER, shuffle=True, random_state=RANDOM_STATE)
            search = GridSearchCV(pipe, grid, cv=inner, scoring=cfg.TUNING_SCORING, n_jobs=-1)
            search.fit(X[c_train], y[c_train])

            proba = search.best_estimator_.predict_proba(X[c_val])[:, 1]
            pred = (proba >= THRESHOLD).astype(int)
            y_val = y[c_val].to_numpy()
            auc = roc_auc_score(y_val, proba) if len(np.unique(y_val)) > 1 else np.nan

            cluster_rows.append({"model": name, "fold": k, "cluster": c,
                                 "n_train": int(c_train.sum()), "n_val": len(y_val),
                                 "accuracy": accuracy_score(y_val, pred),
                                 "f1": f1_score(y_val, pred, zero_division=0), "roc_auc": auc})
            best_param_rows.append({"model": name, "fold": k, "cluster": c, "best_params": search.best_params_})
            group_auc.append(auc)
            group_n.append(len(y_val))
            for pid, yt, p, cl in zip(patient_id[c_val], y_val, proba, pred):
                oof_rows.append({"patient_id": int(pid), "fold": k, "cluster": c, "model": name,
                                 "y_true": int(yt), "proba": float(p), "pred": int(cl)})
            pooled_true.append(y_val)
            pooled_proba.append(proba)
            pooled_pred.append(pred)

        yt_all, pr_all, pd_all = map(np.concatenate, (pooled_true, pooled_proba, pooled_pred))
        aucs, ns = np.array(group_auc), np.array(group_n, dtype=float)
        ok = ~np.isnan(aucs)
        fold_rows.append({
            "model": name, "fold": k,
            "accuracy": accuracy_score(yt_all, pd_all),
            "f1": f1_score(yt_all, pd_all, zero_division=0),
            "roc_auc": roc_auc_score(yt_all, pr_all),
            "roc_auc_within_group": float(np.average(aucs[ok], weights=ns[ok])) if ok.any() else np.nan,
        })

model_order = {m: i for i, m in enumerate(cfg.MODEL_NAMES)}
fold_results_df = pd.DataFrame(fold_rows).sort_values(["model", "fold"], key=lambda s: s.map(model_order) if s.name == "model" else s).reset_index(drop=True)
cluster_results_df = pd.DataFrame(cluster_rows)
oof_df = pd.DataFrame(oof_rows)
pd.DataFrame(best_param_rows).to_csv(os.path.join(cfg.DIAGNOSTICS_DIR, "armC_best_params.csv"), index=False)
cluster_results_df.to_csv(os.path.join(cfg.DIAGNOSTICS_DIR, "armC_cluster_fold_results.csv"), index=False)
print("Nested cross validation complete:", len(fold_results_df), "classifier x fold rows,",
      len(cluster_results_df), "classifier x fold x cluster rows")

Nested cross validation complete: 30 classifier x fold rows, 60 classifier x fold x cluster rows


## 8. Results

The files saved in `results/` match Arms A and B:

* `results/armC_fold_results.csv`: pooled accuracy, F1 and ROC AUC per classifier and fold, plus `roc_auc_within_group`.
* `results/armC_results.csv`: mean and standard deviation across the 10 folds of the pooled metrics.
* `results/predictions/armC_predictions.csv`: one out of fold prediction per patient per classifier, with the cluster the patient was assigned to in that fold.

Cluster level scores, cluster choices, chosen hyperparameters and the descriptive clustering outputs are in `diagnostics/`. Cluster numbers are only meaningful within a fold (k prototypes numbers its clusters afresh each time), so no table averages "cluster 0" across folds. Confusion matrices and tests between arms are in `cross_arm_comparison.ipynb`.

In [11]:
assert oof_df.groupby("model")["patient_id"].nunique().eq(len(df)).all()
assert oof_df.groupby(["model", "patient_id"]).size().max() == 1

fold_results_df.to_csv(os.path.join(cfg.RESULTS_DIR, "armC_fold_results.csv"), index=False)
oof_df.sort_values(["model", "patient_id"], key=lambda s: s.map(model_order) if s.name == "model" else s) \
      .to_csv(os.path.join(cfg.PREDICTIONS_DIR, "armC_predictions.csv"), index=False)

results = fold_results_df.groupby("model", sort=False)[["accuracy", "f1", "roc_auc"]].agg(["mean", "std"])
results.columns = [f"{m}_{s}" for m, s in results.columns]
results = results.reset_index()
results.to_csv(os.path.join(cfg.RESULTS_DIR, "armC_results.csv"), index=False)
results.set_index("model").round(3)

,accuracy_mean,accuracy_std,f1_mean,f1_std,roc_auc_mean,roc_auc_std
model,,,,,,
Logistic regression,0.835,0.055,0.817,0.065,0.894,0.050
Random forest,0.828,0.051,0.805,0.064,0.915,0.044
Support vector machine,0.828,0.043,0.808,0.051,0.900,0.042


## 9. Notes

* **Per fold clustering.** Every fold chose two clusters and no fold fell back to the global model. Every fold reproduces the severity split seen in the whole dataset fit: one cluster with about 80% disease and one with about 20%.
* **Pooled results.** Logistic regression: accuracy 0.835, F1 0.817, ROC AUC 0.894. Random forest: 0.828, 0.805, 0.915. Support vector machine: 0.828, 0.808, 0.900 (means across 10 folds). Whether these differ from Arms A and B is tested in `cross_arm_comparison.ipynb`.
* **ROC AUC caveat.** Because the two clusters have very different disease rates, the pooled ROC AUC can look good simply because patients from the high risk cluster receive higher probabilities than those from the low risk cluster, even if each cluster's model ranks its own patients less well. `roc_auc_within_group` is reported alongside it for that reason, and the comparison notebook shows both.
* **Controlled comparison.** Compared with Arm A, the only difference is that patients are clustered within each training fold and each cluster gets its own tuned model. Data, all 9 features, preprocessing, classifiers, grids, tuning, folds and threshold come from `study_config.py` and `fold_id.csv`, and are checked against `run_manifest.json` before the models run.